File: application/src/test/java/de/seuhd/campuscoffee/tests/SystemTestUtils.java
package de.seuhd.campuscoffee.tests;

import de.seuhd.campuscoffee.api.dtos.PosDto;
import de.seuhd.campuscoffee.api.dtos.ReviewDto;
import de.seuhd.campuscoffee.api.dtos.UserDto;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.web.servlet.client.EntityExchangeResult;
import org.springframework.test.web.servlet.client.RestTestClient;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.utility.DockerImageName;

import java.lang.reflect.Array;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Utility class for system tests.
 */
public class SystemTestUtils {

    /** Client bound to the running server for the current test; set via {@link #configureClient(int)}. */
    private static RestTestClient client;

    /**
     * Binds the shared {@link RestTestClient} to the running server on the given port.
     *
     * @param port the random server port for the current test
     */
    public static void configureClient(int port) {
        client = RestTestClient.bindToServer().baseUrl("http://localhost:" + port).build();
    }

    /**
     * The client bound to the running server, for tests that call endpoints outside the CRUD helpers.
     *
     * @return the configured client
     */
    public static RestTestClient client() {
        return client;
    }

    /**
     * Creates and configures a PostgreSQL testcontainer.
     *
     * @return Configured PostgreSQLContainer instance
     */
    @SuppressWarnings("resource")
    public static PostgreSQLContainer<?> getPostgresContainer() {
        return new PostgreSQLContainer<>(
                DockerImageName.parse("postgres:17-alpine"))
                .withUsername("postgres")
                .withPassword("postgres")
                .withDatabaseName("postgres")
                .withReuse(false);
    }

    /**
     * Configures Spring datasource properties to use the provided PostgreSQL testcontainer.
     *
     * @param registry          DynamicPropertyRegistry to add properties to
     * @param postgresContainer PostgreSQLContainer instance
     */
    public static void configurePostgresContainers (DynamicPropertyRegistry registry, PostgreSQLContainer<?> postgresContainer) {
        registry.add("spring.datasource.url", postgresContainer::getJdbcUrl);
        registry.add("spring.datasource.username", postgresContainer::getUsername);
        registry.add("spring.datasource.password", postgresContainer::getPassword);
    }

    /**
     * Asserts that two objects are equal, ignoring specified fields.
     *
     * @param actual         the actual object
     * @param expected       the expected object
     * @param fieldsToIgnore fields to ignore during comparison
     * @param <T>            the type of the objects being compared
     */
    public static <T> void assertEqualsIgnoringFields(T actual, T expected, String... fieldsToIgnore) {
        assertThat(actual)
                .usingRecursiveComparison()
                .ignoringFields(fieldsToIgnore)
                .isEqualTo(expected);
    }

    /**
     * Asserts that two objects are equal, ignoring timestamp fields (createdAt, updatedAt).
     *
     * @param actual   the actual object
     * @param expected the expected object
     * @param <T>      the type of the objects being compared
     */
    public static <T> void assertEqualsIgnoringTimestamps(T actual, T expected) {
        assertEqualsIgnoringFields(actual, expected, "createdAt", "updatedAt");
    }

    /**
     * Asserts that two objects are equal, ignoring ID and timestamp fields.
     *
     * @param actual   the actual object
     * @param expected the expected object
     * @param <T>      the type of the objects being compared
     */
    public static <T> void assertEqualsIgnoringIdAndTimestamps(T actual, T expected) {
        assertEqualsIgnoringFields(actual, expected, "id", "createdAt", "updatedAt");
    }

    /**
     * Asserts that two collections contain the same elements (in any order), ignoring specified fields.
     *
     * @param actual         the actual collection
     * @param expected       the expected collection
     * @param fieldsToIgnore fields to ignore during comparison
     * @param <T>            the type of elements in the collections
     */
    public static <T> void assertEqualsIgnoringFields(List<T> actual, List<T> expected, String... fieldsToIgnore) {
        assertThat(actual)
                .usingRecursiveFieldByFieldElementComparatorIgnoringFields(fieldsToIgnore)
                .containsExactlyInAnyOrderElementsOf(expected);
    }

    /**
     * Asserts that two collections contain the same elements (in any order), ignoring timestamp fields for
     * each element comparison.
     *
     * @param actual   the actual collection
     * @param expected the expected collection
     * @param <T>      the type of elements in the collections
     */
    public static <T> void assertEqualsIgnoringTimestamps(List<T> actual, List<T> expected) {
        assertEqualsIgnoringFields(actual, expected, "createdAt", "updatedAt");
    }

    /**
     * Generic utility for REST API testing over {@link RestTestClient}. Provides reusable methods for
     * common CRUD operations against the server bound by {@link #configureClient(int)}.
     *
     * @param basePath  The base path of the API endpoint
     * @param dtoClass  The DTO class of the entities being tested
     * @param idGetter  Function to extract the ID from the DTO
     */
    public record Requests<T>(
            String basePath,
            Class<T> dtoClass,
            Function<T, Long> idGetter) {

        /** The DTO body of a response, after asserting the expected status. */
        private T body(RestTestClient.ResponseSpec response, HttpStatus expected) {
            EntityExchangeResult<T> result = response.returnResult(dtoClass);
            assertThat(result.getStatus().value()).isEqualTo(expected.value());
            return result.getResponseBody();
        }

        /** The raw status code of a response, without asserting it. */
        private static int status(RestTestClient.ResponseSpec response) {
            return response.returnResult(byte[].class).getStatus().value();
        }

        /** The list body of a response, deserialized via the DTO array type, after asserting 200. */
        @SuppressWarnings("unchecked")
        private List<T> list(RestTestClient.ResponseSpec response) {
            Class<T[]> arrayType = (Class<T[]>) Array.newInstance(dtoClass, 0).getClass();
            EntityExchangeResult<T[]> result = response.returnResult(arrayType);
            assertThat(result.getStatus().value()).isEqualTo(HttpStatus.OK.value());
            T[] elements = result.getResponseBody();
            return elements == null ? List.of() : List.of(elements);
        }

        /**
         * Retrieves all entities via the API.
         *
         * @return List of DTOs representing all entities
         */
        public List<T> retrieveAll() {
            return list(client.get().uri(basePath).accept(MediaType.APPLICATION_JSON).exchange());
        }

        /**
         * Retrieves an entity by its ID via the API.
         *
         * @param id ID of the entity to retrieve
         * @return DTO representing the retrieved entity
         */
        public T retrieveById(Long id) {
            return body(client.get().uri(basePath + "/{id}", id).accept(MediaType.APPLICATION_JSON).exchange(),
                    HttpStatus.OK);
        }

        /**
         * Retrieves an entity by a filter parameter via the API.
         *
         * @param filterValue Value to filter by
         * @return DTO representing the retrieved entity
         */
        public T retrieveByFilter(String filterParameter, String filterValue) {
            return body(client.get()
                    .uri(basePath + "/filter?" + filterParameter + "={value}", filterValue)
                    .accept(MediaType.APPLICATION_JSON).exchange(), HttpStatus.OK);
        }

        /**
         * Filters by a parameter and returns the raw HTTP status code instead of asserting 200. A test
         * uses this to assert a 404 when no entity matches the filter value.
         *
         * @param filterParameter the filter query parameter name
         * @param filterValue     the value to filter by
         * @return the HTTP status code of the response
         */
        public int retrieveByFilterStatusCode(String filterParameter, String filterValue) {
            return status(client.get()
                    .uri(basePath + "/filter?" + filterParameter + "={value}", filterValue)
                    .accept(MediaType.APPLICATION_JSON).exchange());
        }

        /**
         * Creates multiple entities via the API and returns their DTOs.
         *
         * @param entityList List of DTOs to create
         * @return List of DTOs representing the created entities (including their IDs)
         */
        public List<T> create(List<T> entityList) {
            return entityList.stream()
                    .map(dto -> body(client.post().uri(basePath)
                            .contentType(MediaType.APPLICATION_JSON).body(dto).exchange(), HttpStatus.CREATED))
                    .toList();
        }

        /**
         * Creates entities via the API and returns the creation status codes.
         *
         * @param entityList List of DTOs to create
         * @return List of status codes from the creation requests
         */
        public List<Integer> createAndReturnStatusCodes(List<T> entityList) {
            return entityList.stream()
                    .map(dto -> status(client.post().uri(basePath)
                            .contentType(MediaType.APPLICATION_JSON).body(dto).exchange()))
                    .toList();
        }

        /**
         * Updates multiple entities via the API and returns their updated DTOs.
         *
         * @param entityList List of DTOs to update
         * @return List of DTOs representing the updated entities
         */
        public List<T> update(List<T> entityList) {
            return entityList.stream()
                    .map(dto -> body(client.put().uri(basePath + "/{id}", idGetter.apply(dto))
                            .contentType(MediaType.APPLICATION_JSON).body(dto).exchange(), HttpStatus.OK))
                    .toList();
        }

        /**
         * Deletes multiple entities by their IDs via the API and returns the corresponding status codes.
         *
         * @param idList List of IDs of the entities to delete
         * @return List of HTTP status codes resulting from each delete operation
         */
        public List<Integer> deleteAndReturnStatusCodes(List<Long> idList) {
            return idList.stream()
                    .map(id -> status(client.delete().uri(basePath + "/{id}", id).exchange()))
                    .toList();
        }

        /**
         * Retrieves entities matching several filter query parameters via the API.
         * Unlike {@link #retrieveByFilter(String, String)}, this returns a list (the reviews
         * filter endpoint returns all matching reviews rather than a single entity).
         *
         * @param params query parameter name/value pairs to filter by
         * @return list of DTOs matching the filter criteria
         */
        public List<T> retrieveByFilter(Map<String, Object> params) {
            return list(client.get().uri(uriBuilder -> {
                uriBuilder.path(basePath + "/filter");
                params.forEach(uriBuilder::queryParam);
                return uriBuilder.build();
            }).accept(MediaType.APPLICATION_JSON).exchange());
        }

        /**
         * Retrieves an entity by its ID and returns the raw status code, so a test can assert a 404.
         *
         * @param id ID of the entity to retrieve
         * @return the HTTP status code of the response
         */
        public int retrieveByIdStatusCode(Long id) {
            return status(client.get().uri(basePath + "/{id}", id).accept(MediaType.APPLICATION_JSON).exchange());
        }

        /**
         * Updates an entity using an explicit path ID that may differ from the body ID, returning the
         * raw status code. A test uses this to assert a 400 when the path and body IDs disagree.
         *
         * @param pathId the ID to put in the request path
         * @param dto    the entity body to send
         * @return the HTTP status code of the response
         */
        public int updateWithPathIdAndReturnStatusCode(Long pathId, T dto) {
            return status(client.put().uri(basePath + "/{id}", pathId)
                    .contentType(MediaType.APPLICATION_JSON).body(dto).exchange());
        }

        /**
         * Updates multiple entities via the API and returns the resulting status codes, so a test can
         * assert a 404 when updating an entity that does not exist.
         *
         * @param entityList List of DTOs to update
         * @return List of HTTP status codes resulting from each update operation
         */
        public List<Integer> updateAndReturnStatusCodes(List<T> entityList) {
            return entityList.stream()
                    .map(dto -> status(client.put().uri(basePath + "/{id}", idGetter.apply(dto))
                            .contentType(MediaType.APPLICATION_JSON).body(dto).exchange()))
                    .toList();
        }

        /**
         * Approves an entity on behalf of a user via {@code PUT /{id}/approve?user_id=...} (reviews).
         *
         * @param id     ID of the entity to approve
         * @param userId ID of the user performing the approval
         * @return the updated DTO
         */
        public T approve(Long id, Long userId) {
            return body(client.put().uri(basePath + "/{id}/approve?user_id={userId}", id, userId).exchange(),
                    HttpStatus.OK);
        }

        /**
         * Approves an entity and returns the raw status code, so a test can assert a 400 when a user
         * approves their own review or a 404 when the review does not exist.
         *
         * @param id     ID of the entity to approve
         * @param userId ID of the user performing the approval
         * @return the HTTP status code of the response
         */
        public int approveAndReturnStatusCode(Long id, Long userId) {
            return status(client.put().uri(basePath + "/{id}/approve?user_id={userId}", id, userId).exchange());
        }

        public static SystemTestUtils.Requests<PosDto> posRequests = new SystemTestUtils.Requests<>(
                "/api/pos",
                PosDto.class,
                PosDto::getId
        );

        public static SystemTestUtils.Requests<UserDto> userRequests = new SystemTestUtils.Requests<>(
                "/api/users",
                UserDto.class,
                UserDto::getId
        );

        public static SystemTestUtils.Requests<ReviewDto> reviewRequests = new SystemTestUtils.Requests<>(
                "/api/reviews",
                ReviewDto.class,
                ReviewDto::getId
        );
    }
}

File: application/src/test/kotlin/de/seuhd/campuscoffee/tests/SystemTestUtils.kt
package de.seuhd.campuscoffee.tests

import de.seuhd.campuscoffee.api.dtos.PosDto
import de.seuhd.campuscoffee.api.dtos.ReviewDto
import de.seuhd.campuscoffee.api.dtos.UserDto
import de.seuhd.campuscoffee.domain.model.objects.Role
import de.seuhd.campuscoffee.domain.tests.TestFixtures
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.recursive.comparison.RecursiveComparisonConfiguration
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.test.context.DynamicPropertyRegistry
import org.springframework.test.web.servlet.client.RestTestClient
import org.springframework.test.web.servlet.client.returnResult
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.utility.DockerImageName
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.lang.reflect.Array as ReflectArray

/**
 * Utilities for the system tests.
 */
object SystemTestUtils {
    /**
     * Client bound to the running server for the current test. It is `lateinit` because the value depends
     * on the embedded server's port, which exists only after the server starts, so it is set in
     * [configureClient] rather than at construction.
     */
    private lateinit var client: RestTestClient

    // One client per server port, created once and reused for that context's lifetime. Without fork-level
    // parallelism (-PtestForks=1) every system-test context is cached in a single JVM, each on its own
    // random port; rebuilding the client on every test left a client (with its own connection pool) open
    // per test, hundreds at once, whose crossed connections surfaced as wrong-port I/O errors. Keying by
    // port hands each test the client for its own context's server and bounds the live clients to one per
    // context.
    private val clientsByPort = ConcurrentHashMap<Int, RestTestClient>()

    /**
     * The credentials a request authenticates with (HTTP Basic). The seeded fixture users below carry the
     * roles needed for the write paths; tests that exercise ownership or role rules act as a specific user.
     */
    data class Credentials(
        val loginName: String,
        val password: String
    )

    /** Seeded fixture credentials, derived from [TestFixtures] so each password is defined in one place. */
    val ADMIN = credentialsFor(Role.ADMIN)
    val MODERATOR = credentialsFor(Role.MODERATOR)
    val USER = credentialsFor(Role.USER)

    /** An admin who is not also a moderator, to show content moderation is gated on MODERATOR, not ADMIN. */
    val ADMIN_NO_MOD =
        TestFixtures.adminWithoutModeration().let {
            Credentials(
                requireNotNull(it.loginName),
                requireNotNull(it.password)
            )
        }

    private fun credentialsFor(role: Role): Credentials =
        TestFixtures.rawCredentialsFor(role).let { (login, password) -> Credentials(login, password) }

    /**
     * The credentials the CRUD helpers and [client] authenticate with by default. Defaults to the admin
     * fixture (full roles), so a write request succeeds unless a test narrows the actor. Reset before each test by
     * the base class so a per-test override does not leak.
     */
    var defaultCredentials: Credentials = ADMIN

    /** Selects the [RestTestClient] for the running server on the given port, creating it once per port. */
    fun configureClient(port: Int) {
        client =
            clientsByPort.computeIfAbsent(port) {
                RestTestClient.bindToServer().baseUrl("http://localhost:$it").build()
            }
        defaultCredentials = ADMIN
    }

    /** The Basic-auth header value for the given credentials, as Spring's [HttpHeaders.setBasicAuth] builds it. */
    fun basicAuthHeader(credentials: Credentials): String =
        HttpHeaders()
            .apply { setBasicAuth(credentials.loginName, credentials.password) }
            .getFirst(HttpHeaders.AUTHORIZATION)!!

    /** The client bound to the running server, for tests that call endpoints outside the CRUD helpers. */
    fun client(): RestTestClient = client

    // Creates a PostgreSQL testcontainer. The container is AutoCloseable but deliberately not closed here:
    // callers keep it open for the whole test run and Testcontainers tears it down on JVM shutdown, so
    // suppress the resource leak inspection.
    @Suppress("resource")
    fun getPostgresContainer(): PostgreSQLContainer<*> =
        // PostgreSQLContainer is a Java class whose type parameter refers back to itself
        // (PostgreSQLContainer<SELF extends PostgreSQLContainer<SELF>>). Testcontainers uses this so the
        // fluent withX() setters return the concrete subclass type, letting them chain across the class
        // hierarchy. Kotlin cannot express that self-reference, so we pass Nothing as the type argument; this
        // is the standard way to use Testcontainers from Kotlin. We don't set a custom username/password here
        // and rely on Testcontainers' defaults, which configurePostgresContainers() (below) hands to Spring as
        // the datasource URL, username, and password.
        PostgreSQLContainer<Nothing>(DockerImageName.parse("postgres:18-alpine"))

    /** Points the Spring datasource at the given PostgreSQL testcontainer. */
    fun configurePostgresContainers(
        registry: DynamicPropertyRegistry,
        postgresContainer: PostgreSQLContainer<*>
    ) {
        registry.add("spring.datasource.url", postgresContainer::getJdbcUrl)
        registry.add("spring.datasource.username", postgresContainer::getUsername)
        registry.add("spring.datasource.password", postgresContainer::getPassword)
    }

    /** Asserts two objects are equal, ignoring the given fields. */
    fun <T> assertEqualsIgnoringFields(
        actual: T,
        expected: T,
        vararg fieldsToIgnore: String
    ) {
        assertThat(actual)
            .usingRecursiveComparison()
            .ignoringFields(*fieldsToIgnore)
            // a user's role set has no order; the read path yields an EnumSet, the fixtures a LinkedHashSet
            .ignoringCollectionOrder()
            .isEqualTo(expected)
    }

    // Server-managed fields that never match a client-supplied fixture and so are ignored when comparing
    // business field values: the timestamps and optimistic-locking version the server assigns, and a user's
    // secrets (the raw password is write-only and the stored hash is never serialized).
    private val serverManagedFields = arrayOf("createdAt", "updatedAt", "version", "password", "passwordHash")

    /** Asserts two objects are equal, ignoring the server-managed fields (timestamps, version, user secrets). */
    fun <T> assertEqualsIgnoringTimestamps(
        actual: T,
        expected: T
    ) = assertEqualsIgnoringFields(actual, expected, *serverManagedFields)

    /** Asserts two objects are equal, ignoring the id and the server-managed fields (timestamps, version, secrets). */
    fun <T> assertEqualsIgnoringIdAndTimestamps(
        actual: T,
        expected: T
    ) = assertEqualsIgnoringFields(actual, expected, "id", *serverManagedFields)

    /** Asserts two lists contain the same elements (any order), ignoring the given fields per element. */
    fun <T> assertEqualsIgnoringFields(
        actual: List<T>,
        expected: List<T>,
        vararg fieldsToIgnore: String
    ) {
        val config =
            RecursiveComparisonConfiguration
                .builder()
                .withIgnoredFields(*fieldsToIgnore)
                // a user's role set has no order; the read path yields an EnumSet, the fixtures a LinkedHashSet
                .withIgnoreCollectionOrder(true)
                .build()
        assertThat(actual)
            .usingRecursiveFieldByFieldElementComparator(config)
            .containsExactlyInAnyOrderElementsOf(expected)
    }

    /** Asserts two lists contain the same elements (any order), ignoring the server-managed fields. */
    fun <T> assertEqualsIgnoringTimestamps(
        actual: List<T>,
        expected: List<T>
    ) = assertEqualsIgnoringFields(actual, expected, *serverManagedFields)

    /**
     * Reusable CRUD operations over [RestTestClient] against the server bound by [configureClient].
     *
     * @param basePath the base path of the API endpoint
     * @param dtoClass the DTO class of the entities being tested
     * @param idGetter extracts the id from a DTO
     * @param requestBody the value sent as the request body for a DTO; defaults to the DTO itself. The
     *   user requests override it so the write-only password reaches the server, which the client would
     *   otherwise drop when it serializes the DTO (see [userRequests]).
     */
    class Requests<T : Any>(
        private val basePath: String,
        private val dtoClass: Class<T>,
        private val idGetter: (T) -> UUID?,
        private val requestBody: (T) -> Any = { it }
    ) {
        /** The DTO body of a response, after asserting the expected status. */
        private fun body(
            response: RestTestClient.ResponseSpec,
            expected: HttpStatus
        ): T {
            val result = response.returnResult(dtoClass)
            assertThat(result.status.value()).isEqualTo(expected.value())
            return result.responseBody!!
        }

        /** The raw status code of a response, without asserting it. */
        private fun status(response: RestTestClient.ResponseSpec): Int =
            response.returnResult<ByteArray>().status.value()

        /** The list body of a response, deserialized via the DTO array type, after asserting 200. */
        @Suppress("UNCHECKED_CAST")
        private fun list(response: RestTestClient.ResponseSpec): List<T> {
            val arrayType = ReflectArray.newInstance(dtoClass, 0).javaClass as Class<Array<T>>
            val result = response.returnResult(arrayType)
            assertThat(result.status.value()).isEqualTo(HttpStatus.OK.value())
            return result.responseBody?.toList() ?: emptyList()
        }

        // Reads are public for POS/reviews (credentials = null); user reads need authentication, so the
        // user tests pass credentials, which adds a Basic-auth header to the request.
        private fun RestTestClient.RequestHeadersSpec<*>.withOptionalAuth(credentials: Credentials?) =
            apply { credentials?.let { header(HttpHeaders.AUTHORIZATION, basicAuthHeader(it)) } }

        fun retrieveAll(credentials: Credentials? = null): List<T> =
            list(
                client
                    .get()
                    .uri(basePath)
                    .accept(MediaType.APPLICATION_JSON)
                    .withOptionalAuth(credentials)
                    .exchange()
            )

        /** Lists all and returns the raw status code (to assert a 401/403 when the read is not public). */
        fun retrieveAllStatusCode(credentials: Credentials? = null): Int =
            status(
                client
                    .get()
                    .uri(basePath)
                    .accept(MediaType.APPLICATION_JSON)
                    .withOptionalAuth(credentials)
                    .exchange()
            )

        fun retrieveById(
            id: UUID,
            credentials: Credentials? = null
        ): T =
            body(
                client
                    .get()
                    .uri("$basePath/{id}", id)
                    .accept(MediaType.APPLICATION_JSON)
                    .withOptionalAuth(credentials)
                    .exchange(),
                HttpStatus.OK
            )

        fun retrieveByFilter(
            filterParameter: String,
            filterValue: String,
            credentials: Credentials? = null
        ): T =
            body(
                client
                    .get()
                    .uri("$basePath/filter?$filterParameter={value}", filterValue)
                    .accept(MediaType.APPLICATION_JSON)
                    .withOptionalAuth(credentials)
                    .exchange(),
                HttpStatus.OK
            )

        /** Filters by a parameter and returns the raw status code (to assert a 404 on a filter miss, or 401/403). */
        fun retrieveByFilterStatusCode(
            filterParameter: String,
            filterValue: String,
            credentials: Credentials? = null
        ): Int =
            status(
                client
                    .get()
                    .uri("$basePath/filter?$filterParameter={value}", filterValue)
                    .accept(MediaType.APPLICATION_JSON)
                    .withOptionalAuth(credentials)
                    .exchange()
            )

        fun create(
            entityList: List<T>,
            credentials: Credentials = defaultCredentials
        ): List<T> =
            entityList.map { dto ->
                body(
                    client
                        .post()
                        .uri(basePath)
                        .header(HttpHeaders.AUTHORIZATION, basicAuthHeader(credentials))
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(requestBody(dto))
                        .exchange(),
                    HttpStatus.CREATED
                )
            }

        fun createAndReturnStatusCodes(
            entityList: List<T>,
            credentials: Credentials = defaultCredentials
        ): List<Int> =
            entityList.map { dto ->
                status(
                    client
                        .post()
                        .uri(basePath)
                        .header(HttpHeaders.AUTHORIZATION, basicAuthHeader(credentials))
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(requestBody(dto))
                        .exchange()
                )
            }

        /** Creates without an Authorization header (to assert a 401 on an unauthenticated write request). */
        fun createUnauthenticatedAndReturnStatusCode(dto: T): Int =
            status(
                client
                    .post()
                    .uri(basePath)
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(requestBody(dto))
                    .exchange()
            )

        fun update(
            entityList: List<T>,
            credentials: Credentials = defaultCredentials
        ): List<T> =
            entityList.map { dto ->
                body(
                    client
                        .put()
                        .uri("$basePath/{id}", idGetter(dto))
                        .header(HttpHeaders.AUTHORIZATION, basicAuthHeader(credentials))
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(requestBody(dto))
                        .exchange(),
                    HttpStatus.OK
                )
            }

        fun deleteAndReturnStatusCodes(
            idList: List<UUID>,
            credentials: Credentials = defaultCredentials
        ): List<Int> =
            idList.map { id ->
                status(
                    client
                        .delete()
                        .uri("$basePath/{id}", id)
                        .header(HttpHeaders.AUTHORIZATION, basicAuthHeader(credentials))
                        .exchange()
                )
            }

        /** Filters by several query parameters, returning a list (the reviews filter returns many). */
        fun retrieveByFilter(params: Map<String, Any>): List<T> =
            list(
                client
                    .get()
                    .uri { uriBuilder ->
                        uriBuilder.path("$basePath/filter")
                        params.forEach { (key, value) -> uriBuilder.queryParam(key, value) }
                        uriBuilder.build()
                    }.accept(MediaType.APPLICATION_JSON)
                    .exchange()
            )

        /** Retrieves by id and returns the raw status code (to assert a 404, or a 401/403 on a guarded read). */
        fun retrieveByIdStatusCode(
            id: UUID,
            credentials: Credentials? = null
        ): Int =
            status(
                client
                    .get()
                    .uri("$basePath/{id}", id)
                    .accept(MediaType.APPLICATION_JSON)
                    .withOptionalAuth(credentials)
                    .exchange()
            )

        /** Updates with an explicit path id that may differ from the body id (to assert a 400 on mismatch). */
        fun updateWithPathIdAndReturnStatusCode(
            pathId: UUID,
            dto: T,
            credentials: Credentials = defaultCredentials
        ): Int =
            status(
                client
                    .put()
                    .uri("$basePath/{id}", pathId)
                    .header(HttpHeaders.AUTHORIZATION, basicAuthHeader(credentials))
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(requestBody(dto))
                    .exchange()
            )

        /** Updates and returns the status codes (to assert a 404 when updating a missing entity, or a 403). */
        fun updateAndReturnStatusCodes(
            entityList: List<T>,
            credentials: Credentials = defaultCredentials
        ): List<Int> =
            entityList.map { dto ->
                status(
                    client
                        .put()
                        .uri("$basePath/{id}", idGetter(dto))
                        .header(HttpHeaders.AUTHORIZATION, basicAuthHeader(credentials))
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(requestBody(dto))
                        .exchange()
                )
            }

        /** Approves an entity as the authenticated user via PUT /{id}/approve (the approver is the caller). */
        fun approve(
            id: UUID,
            credentials: Credentials = defaultCredentials
        ): T =
            body(
                client
                    .put()
                    .uri("$basePath/{id}/approve", id)
                    .header(HttpHeaders.AUTHORIZATION, basicAuthHeader(credentials))
                    .exchange(),
                HttpStatus.OK
            )

        /** Approves and returns the raw status code (to assert a 400 self-approval, 404, or 409 repeat). */
        fun approveAndReturnStatusCode(
            id: UUID,
            credentials: Credentials = defaultCredentials
        ): Int =
            status(
                client
                    .put()
                    .uri("$basePath/{id}/approve", id)
                    .header(HttpHeaders.AUTHORIZATION, basicAuthHeader(credentials))
                    .exchange()
            )

        /** Approves with a raw bearer token (to assert a JWT-authenticated write request or an expired/forged 401). */
        fun approveWithBearerAndReturnStatusCode(
            id: UUID,
            token: String
        ): Int =
            status(
                client
                    .put()
                    .uri("$basePath/{id}/approve", id)
                    .header(HttpHeaders.AUTHORIZATION, "Bearer $token")
                    .exchange()
            )

        /** Creates with a raw bearer token (to assert that a JWT-authenticated write request obeys the same rules). */
        fun createWithBearerAndReturnStatusCode(
            dto: T,
            token: String
        ): Int =
            status(
                client
                    .post()
                    .uri(basePath)
                    .header(HttpHeaders.AUTHORIZATION, "Bearer $token")
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(requestBody(dto))
                    .exchange()
            )

        /** Reverts the last change via POST /{id}/revert?version=, returning the restored DTO after asserting 200. */
        fun revert(
            id: UUID,
            version: Long,
            credentials: Credentials = defaultCredentials
        ): T =
            body(
                client
                    .post()
                    .uri("$basePath/{id}/revert?version={version}", id, version)
                    .header(HttpHeaders.AUTHORIZATION, basicAuthHeader(credentials))
                    .exchange(),
                HttpStatus.OK
            )

        /** Reverts and returns the raw status code (to assert a 204, 400, 401, 403, 404, or 409). */
        fun revertAndReturnStatusCode(
            id: UUID,
            version: Long,
            credentials: Credentials? = defaultCredentials
        ): Int =
            status(
                client
                    .post()
                    .uri("$basePath/{id}/revert?version={version}", id, version)
                    .withOptionalAuth(credentials)
                    .exchange()
            )
    }

    val posRequests = Requests("/api/pos", PosDto::class.java, idGetter = { it.id })

    // A UserDto's password is write-only, so the client drops it when it serializes the DTO. The user
    // requests send a map that carries the password instead, which is the body a real client would post.
    val userRequests =
        Requests("/api/users", UserDto::class.java, idGetter = { it.id }, requestBody = ::userRequestBody)
    val reviewRequests = Requests("/api/reviews", ReviewDto::class.java, idGetter = { it.id })

    /**
     * The request body for writing a [UserDto]. The write-only [UserDto.password] survives serialization
     * only when it is sent as a plain map entry; a field a request omits is left out so the server's
     * "omitted means unchanged" rule (on update) and its validation (on create) see the same body a real
     * client would send.
     */
    private fun userRequestBody(dto: UserDto): Map<String, Any?> =
        buildMap {
            dto.id?.let { put("id", it) }
            put("loginName", dto.loginName)
            put("emailAddress", dto.emailAddress)
            put("firstName", dto.firstName)
            put("lastName", dto.lastName)
            dto.password?.let { put("password", it) }
            dto.roles?.let { put("roles", it) }
        }
}
